subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/SelectiveCompiler.java

```diff
@@ -88,20 +91,17 @@ public WorkResult execute(T spec) {
             return rebuildAllCompiler.execute(spec);
         }
 
-        boolean cleanedOutput = recompilationSpecProvider.initializeCompilation(spec, recompilationSpec);
-
-        if (Iterables.isEmpty(spec.getSourceFiles()) && spec.getClasses().isEmpty()) {
-            LOG.info("None of the classes needs to be compiled! Analysis took {}. ", clock.getElapsed());
-            return new RecompilationNotNecessary(previousCompilationData, recompilationSpec);
-        }
-
-        try {
+        CompileTransaction transaction = recompilationSpecProvider.initCompilationSpecAndTransaction(spec, recompilationSpec);
+        return transaction.execute(workResult -> {
+            if (Iterables.isEmpty(spec.getSourceFiles()) && spec.getClasses().isEmpty()) {
+                LOG.info("None of the classes needs to be compiled! Analysis took {}. ", clock.getElapsed());
+                return new RecompilationNotNecessary(previousCompilationData, recompilationSpec);
+            }
             WorkResult result = recompilationSpecProvider.decorateResult(recompilationSpec, previousCompilationData, cleaningCompiler.getCompiler().execute(spec));
-            return result.or(WorkResults.didWork(cleanedOutput));
-        } finally {
             Collection<String> classesToCompile = recompilationSpec.getClassesToCompile();
             LOG.info("Incremental compilation of {} classes completed in {}.", classesToCompile.size(), clock.getElapsed());
             LOG.debug("Recompiled classes {}", classesToCompile);
-        }
```

Comment:
I removed that, since I am not sure what is the added value in case of a failure. If compilation fails, information about recompiled classes is not accurate (normally 0 classes will get recompiled). We can print how many classes we want to compile before a try if that is useful information.

Reply:
Ok, I can readd it. Yes `classesToCompile` is independent. But then `Recompiled classes <some number>` is not actually true, since  compilation failed hehe. But that's not that important.